### PR TITLE
Vastly optimize video decoding performance of high-res (4K) files

### PR DESCRIPTION
**Pull request by lukasf**
- Eliminate sws_scaler where possible (MF will do scaling on GPU)
- Use Platform::ArrayReference to avoid unneccessary copy
- Copy sws_scaler output in one operation (contiguous buffer)

### CODE REVIEW

**Comment by lukasf** [comment]
See discussion in #202. With this patch, I am able to software decode 4K HEVC files 100% smooth, which were previously absolutely unplayable. Even 4K Hi10P files play well. Performance for HD videos has also improved (lower CPU usage).

**Comment by mcosmin222** [comment]
Looks good.

**Comment by reego-fr** [comment]
Looks good but apparently yuvj420 and yuv420 store levels differently. Adding videoProperties->Properties->Insert(MF_MT_VIDEO_NOMINAL_RANGE, (uint32)MFNominalRange_0_255); prevents clipping in the highlights and shadows when the decoder output is yuvj420. 
Another thing non related to these changes but I noticed while testing a yuvj420 file is that the colors seem a bit different when using MF video decoding (h264) compared to VLC. Did you notice anything similar ?

**Comment by lukasf** [comment]
Thanks for the hint with the levels! I originally compared my output with MPC-HC and found no difference in levels so I thought they were correct. But it looks like MPC-HC also does not treat levels correctly. Only when comparing with other video apps now, I saw that levels are really off. I updated the PR as you said, this fixes the levels. The old solutio with sws_scaler also had the levels wrong, looks like it does not update the levels when changing pix format.

And yes I also noticed already that colors are slightly off for HEVC files! Can h264 files have YUVJ as well? I only saw that pix format on HEVC files. To me it looks like green is slightly oversaturated while red is slightly undersaturated. Not sure where this comes from. Tried playing with MF_MT_YUV_MATRIX but it did not have any effect.

**Comment by lukasf** [comment]
After checking the colors again with the levels fixed, I do not see much of a difference anymore. There is still a very slight difference, but it is even hard to tell what the difference is. I was not able to get rid of it, and did not find any help on google. For me, the difference is too small right now to further look into this, but feel free to open an issue if it is a problem for you.

**Comment by reego-fr** [comment]
I will try to find more samples files to try to figure out what can be wrong, I am not even sure VLC output is what my sample should reallly look like. 
It is apparently the transfer characteristic of "gamma2.2" in my file that introduce this inconsistency but I could not see any change when playing with MF_MT_TRANSFER_FUNCTION attribute. 
I will probably open a new issue once I get more testing material.

**Comment by mcosmin222** [comment]
Try playing the file in windows media player classic with ffmpeg codec pack. It should give you a good idea about the colors.

**Comment by reego-fr** [comment]
I can not reproduce the blatent color shift I am seeing on another machine, it's probably not a big deal considering my weird test setup. 
Anyway, as I understand the color matrix or other parameters should influence yuv<->rgb conversion, so I think one time or another we will have to care about that in the case of uncompressed video playback.

**Comment by reego-fr** [comment]
One last thing about Iuyv uncompressed support. Did you test it under Windows 8.1 Phone and desktop ? I wondered about retrocompatibility but just realized I could not build those projects anymore and test that.

**Comment by mcosmin222** [comment]
At this point I don't think backwards compatibility should be an issue anymore. Perhaps there could be a branch for windows (phone) 8.1 specifically. Microsoft abandonment of these platforms makes it hard for us to use new media features in newer OS version while keeping it compatible with the older versions.

I've personally stopped updating the WP8.1 ffmpeg interop in my app.

**Comment by lukasf** [comment]
When implementing this, I checked MSDN docs for platform support of these formats. The output formats should be supported on all platforms. Only limitation is RGBA output on WP8.1, where the alpha channel will be dropped during playback (but it should still render).

I agree with @mcosmin222 about backwards compatibility. I already planned to submit an issue to remove support for Windows 8 and WP 8 and update the project files to VS2017 toolkit. A branch can be created for legacy apps. I don't think that anyone here really cares about Windows 8 or even WP 8, and no one would start new apps for those platforms.

**Comment by lukasf** [comment]
I just created #206 for dropping W(P)8.1 support. Please post your opinions there.

File: FFmpegInterop/Source/UncompressedVideoSampleProvider.cpp [hunk 1]
```diff
@@ -27,18 +27,29 @@ extern "C"
...
 	{
-		m_rgVideoBufferLineSize[i] = 0;
-		m_rgVideoBufferData[i] = nullptr;
+	case AV_PIX_FMT_YUV420P:
+	case AV_PIX_FMT_YUVJ420P:
+		m_OutputPixelFormat = AV_PIX_FMT_YUV420P;
+		OutputMediaSubtype = MediaEncodingSubtypes::Iyuv;
+		break;
+	case AV_PIX_FMT_YUVA420P:
+		m_OutputPixelFormat = AV_PIX_FMT_BGRA;
```
Comment (reego-fr) [inline FFmpegInterop/Source/UncompressedVideoSampleProvider.cpp]: Isn't there a typo on AV_PIX_FMT_BGRA ?

Comment (lukasf) [reply FFmpegInterop/Source/UncompressedVideoSampleProvider.cpp]: Nope, ffmpeg BGRA = MF ARGB

Comment (reego-fr) [reply FFmpegInterop/Source/UncompressedVideoSampleProvider.cpp]: Ok, then it is AVPixelFormat::AV_PIX_FMT_RGBA in WriteAVPacketToStream that's wrong no ?

Comment (lukasf) [reply FFmpegInterop/Source/UncompressedVideoSampleProvider.cpp]: Fixed, thanks. It still worked (by chance) because LineSize[1] is 0 for BGRA...

**Comment by mcosmin222** [comment]
@khouzam , Maybe it is time for the merge?

**Comment by khouzam** [comment]
Let me find a bit of time to go over the code.

**Comment by lukasf** [comment]
I accidentially pushed a wrong changeset into this branch. Reverted and merged latest master branch.

**Comment by khouzam** [comment]
Thank you, I was going to say, let's get the frame grabber feature in a separate branch 😊

**Comment by lukasf** [comment]
Yes, I really can't tell how it ended up here, just found out by coincidence :)

**Comment by reego-fr** [comment]
Hi,
I actually found there is a small issue with direct yuv420p playback. I found files that decode to planes larger than the stream width (i.e. linesize is larger than width). 
On the media foundation side I tried setting MF_MT_DEFAULT_STRIDE attribute on the sample or on the videoStreamDescriptor but it didn't look like it was doing anything.

**Comment by lukasf** [comment]
@reego-fr Can you upload a file somehere?

**Comment by reego-fr** [comment]
Here is one https://1drv.ms/u/s!Av6YXBDKaOdvgutq2u-I2SIafHaTrw 
I don't know if it is limited to non square pixel aspect ratio files but it looks related.

**Comment by reego-fr** [comment]
It happens on 1:1 files too. It is really only dependent on the decoded frame width.

**Comment by lukasf** [comment]
Huh, well, I am also a bit lost here. I tried various ways, not only with MF_MT_DEFAULT_STRIDE but also MF_MT_MINIMUM_DISPLAY_APERTURE and MF_MT_GEOMETRIC_APERTURE. All those don't have any effect at all on the output.

@khouzam Do you think you can get someone from the media team look into that? I also created a post on the msdn forums:

https://social.msdn.microsoft.com/Forums/en-US/6247bc89-f741-4203-80d8-a0db07e3fa0d/uwp-win10-mediastreamsource-setting-mfmtdefaultstride-does-not-have-any-effect?forum=wpdevelop

**Comment by lukasf** [comment]
I checked in a workaround for this problem. If the decoder wants to add stride, I now force use of sws_scale. This seems to work well so far. But of cause it is only a workaround. Best would be if we could use the stride MF attribute.
@reego-fr Can you test the latest update with your files as well?

**Comment by mcosmin222** [comment]
If this is indeed a bug, it will probably be fixed in the next major windows update, so your workaround is there to stay.

**Comment by lukasf** [comment]
Yes I guess so.

**Comment by reego-fr** [comment]
It works for me too. Thanks.

**Comment by lukasf** [comment]
I found out how to use MF_MT_MINIMUM_DISPLAY_APERTURE to work around MF_MT_DEFAULT_STRIDE not working. Latest checkin does the trick!

**Comment by mcosmin222** [comment]
I think there was a misunderstanding with the default stride actually does. The documentation is a bit misleading. The default stride should be the actual size of the video frame. I this case, 720 not 768. Msdn mentions that default stride only has effect when there are no padding bytes. I think this attribute is intended for special case video frames and that your solution is the more generic one, not the other way around.

**Comment by lukasf** [comment]
Ah, the MS Media guy also told me to use default stride of 768, but he also recommended setting aperture as well. That is why I tried again with the latter and got it working in the end...

**Comment by reego-fr** [comment]
It works for me

**Comment by lukasf** [comment]
Hey guys, I did not really think that I could get this working so quickly: Right now I already have a solution running on a different branch, which totally elimitates any buffer copying. When no scaler is used, the decoder directly decodes into what will be the target IBuffer. When scaler is used, the scaler will render into the target IBuffer. Plus, BufferPools are used for both solutions, to avoid unneccessary malloc/free, this helps keep memory consumption stable. No more memory copy at all, no DataWriter, this is as efficient as it can get.

If you have some time, please check out the branch "lukasf/experimental-direct-buffer", and let me know how it works for you. It already includes all other fixes which were done here (stride problem, color range, etc).

Not sure how we should proceed. Shall we keep this PR, or directly switch to the new, more advanced solution? It still needs some more testing, and some cleanup is needed. But for me, it works pretty well already.

**Comment by mcosmin222** [comment]
I will test it asap. Meanwhile, have you tried a long running file to make sure there are no memory leaks?

**Comment by lukasf** [comment]
Memory consumption is rock stable from what I see.

**Comment by reego-fr** [comment]
It's really nice. No playback issues to report so far.

**Comment by lukasf** [comment]
@reego-fr Cool, that's good to hear!

I have cleaned up the code and made a new branch lukasf/direct-buffer-clean. I also opened a new PR #216 so let's discuss the new solution over there. Maybe we can close this PR if we decide to jump to direct buffer approach.

**Pull request closed without merging**